### PR TITLE
Pretty-print many more constructs and improve formatting

### PR DESCRIPTION
**Pull request by ianbollinger**
There's a lot to review, so let me know if you need me to rebase/squash any of this.
## Implements pretty-printing
- Type classes
- Fixity declarations
- Guards in equations
- Do notation let bindings
- Derived instances
- Values with type annotations `(like :: This)`
- Guards for case expressions
- Foreign data declarations
- Kinds: row, function, and symbol
- Kinded types
- Constrained types
- Positioned declaration refs (comments can now appear inside export/import lists)
- Type operator refs (`type (~>)` in an export/import list)
- Positioned values (things with comments)
- Typed binders
- Positioned binders
## Fixes pretty-printing
- Case expressions
- Type class instance declarations with constraints
- Data types with no constructors (`data Void`)
## Improves formatting
- Sort imports alphabetically, with Prelude first.
- Selectively printing new lines for top-level declarations
- Indenting `where` bindings
- Formatting module declarations like in the core libraries
- Ensuring files end with a new line
- Indenting by 2 spaces by default. This style is utilized consistently and unambiguously by the core libraries. Don't blame me! :)
- Changing the default width in characters to 80. I confirmed this with paf31.
- Removing the space between exported `Constructor` and `(..)`; core libraries use no space.
- Removing the extra space after data type declarations with no arguments.
- Removing the newline after a function definition's equal sign, unless the function contains a case expression.
- Formatting lists and records across lines if they are too long
- Format ADTs with multiple constructors across multiple lines.
- Wrapping long lines
  - If expressions
  - Case expressions
  - Types
  - Function/value definitions
## Miscellaneous
- Eliminate compiler and HLint warnings.
- Remove unused dependencies from `purescript.cabal`.
- Print to standard output if `--output` is not specified.
- Print error messages to standard output and exit with error code.
## Not done yet
- Is it `{ a: 1, b: 2, c: 3 }` or is `{a: 1, b: 2, c: 3}` acceptable?
- An extra space appears before records/row types.
- An extra space appears before case expressions.
- A comment before a do expression can break indentation. (Syntax error!)
- Let bindings are converted to where bindings (the parser will need to be changed to fix this.) This can lead to syntax errors.
- Some comments are lost (another parser change.)
- Long series of function applications should be formatted like:

``` purescript
function
  argument1
  argument2
  etc
```

### CODE REVIEW

File: psc-format/Config.hs [hunk 1]
```diff
@@ -3,4 +3,4 @@ module Config where
 import Prelude (Int)
 
 indentationLevel :: Int
-indentationLevel = 4
\ No newline at end of file
+indentationLevel = 2
```
Comment (archaeron) [inline psc-format/Config.hs:6]: and I thought the PR would be perfect ;)

**Comment by archaeron** [comment]
Looks good. But I think we'll soon need some options :)

It wold probably be good if you could squash your commits- I should squash mines as well.

File: psc-format/Config.hs (hunk 1 above)
Comment (ianbollinger) [inline psc-format/Config.hs:6]: Hah. I prefer 4 spaces myself.

**Comment by ianbollinger** [comment]
Okay, hopefully 10 commits is more manageable than 46. :)

**Pull request merged**